File: Server/PluginSDK/TS/src/actions.ts
/** 提供函数调用式动作构造器，避免作者手写 JSON Patch。 */

import type { ExtensionAction, JsonValue } from "./model.js";
import { Event } from "./model.js";

/** 构造标准动作并统一填充空字段；内部使用避免不同构造器产生结构漂移。 */
function createAction(event: Event, action: ExtensionAction["action"], options: Partial<ExtensionAction> = {}): ExtensionAction {
  return { eventId: event.id, action, patch: options.patch ?? [], annotations: options.annotations ?? [], output: options.output ?? null };
}

/** 校验终止原因和目标主机包含可见字符；纯空白输入会在进入 Host 前报告作者错误。 */
function requireNonBlank(value: string, message: string): string {
  if (value.trim().length === 0) throw new Error(message);
  return value;
}

/** 保持事件不变；适用于全部阶段。 */
export function continueEvent(event: Event): ExtensionAction { return createAction(event, "continue"); }

/** 原子替换 payload；Host 会在写入线路前复验阶段结构。 */
export function modifyPayload(event: Event, payload: JsonValue): ExtensionAction {
  return createAction(event, "modify", { patch: [{ op: "replace", path: "", value: payload }] });
}

/** 替换当前线上字节并保留其他 payload 字段。 */
export function modifyBytes(event: Event, bytes: Uint8Array): ExtensionAction {
  return modifyPayload(event, { ...event.payloadObject(), bytes: [...bytes] });
}

/** 暂存当前流块以等待后续半帧。 */
export function hold(event: Event): ExtensionAction { return createAction(event, "hold"); }
/** 丢弃当前块或事务。 */
export function drop(event: Event): ExtensionAction { return createAction(event, "drop"); }
/** 拒绝当前操作并附带稳定原因。 */
export function reject(event: Event, reason = "pluginRejected"): ExtensionAction {
  return createAction(event, "reject", { output: { reason: requireNonBlank(reason, "拒绝原因不能为空") } });
}
/** 请求关闭当前连接。 */
export function close(event: Event, reason = "pluginClosed"): ExtensionAction {
  return createAction(event, "close", { output: { reason: requireNonBlank(reason, "关闭原因不能为空") } });
}
/** 附加解码树、标签或显示字段，不改变线上数据。 */
export function annotate(event: Event, ...annotations: Record<string, JsonValue>[]): ExtensionAction { return createAction(event, "annotate", { annotations }); }

/** 改写最终目标；非法端口在进入 Host 前失败。 */
export function redirect(event: Event, host: string, port: number): ExtensionAction {
  if (!Number.isInteger(port) || port < 1 || port > 65535) throw new Error("重定向目标必须包含有效主机和端口");
  return createAction(event, "redirect", { output: { host: requireNonBlank(host, "重定向主机不能为空"), port } });
}

/** 生成 HTTP、DNS 或命令阶段的完整合成响应。 */
export function respond(event: Event, response: JsonValue): ExtensionAction { return createAction(event, "respond", { output: response }); }

File: Server/PluginSDK/Python/src/capturePluginSdk/actions.py
"""提供函数调用式动作构造器，插件作者不需要手写 JSON Patch。"""

from __future__ import annotations

from typing import Any

from .model import Action, ActionKind, Event


def requireNonBlank(value: str, message: str) -> str:
    """校验终止原因和目标主机包含可见字符；纯空白输入会在进入宿主前报告作者错误。"""
    if not isinstance(value, str) or not value.strip():
        raise ValueError(message)
    return value


def continueEvent(event: Event) -> Action:
    """保持当前事件不变；适用于所有阶段且不会物化正文副本。"""
    return Action(event.eventId, ActionKind.CONTINUE)


def modifyPayload(event: Event, payload: Any) -> Action:
    """原子替换当前 payload；宿主会在发布线上数据前复验阶段与字节结构。"""
    return Action(event.eventId, ActionKind.MODIFY, [{"op": "replace", "path": "", "value": payload}])


def modifyBytes(event: Event, content: bytes | bytearray | memoryview) -> Action:
    """替换当前数据块并保留其他 payload 字段；用于 TCP、UDP、WebSocket 与正文块。"""
    if not isinstance(event.payload, dict):
        raise ValueError("二进制事件 payload 必须是对象")
    payload = dict(event.payload)
    payload["bytes"] = list(bytes(content))
    return modifyPayload(event, payload)


def hold(event: Event) -> Action:
    """暂存流式事件，等待后续字节；仅应在宿主声明支持 hold 的阶段返回。"""
    return Action(event.eventId, ActionKind.HOLD)


def drop(event: Event) -> Action:
    """丢弃当前数据块或录制事务；实际语义由当前阶段确定。"""
    return Action(event.eventId, ActionKind.DROP)


def reject(event: Event, reason: str = "pluginRejected") -> Action:
    """拒绝当前操作并附带稳定原因；宿主负责映射为对应协议的终止行为。"""
    return Action(event.eventId, ActionKind.REJECT, output={"reason": requireNonBlank(reason, "拒绝原因不能为空")})


def close(event: Event, reason: str = "pluginClosed") -> Action:
    """请求立即关闭当前连接；非连接阶段使用时将由宿主结构校验拒绝。"""
    return Action(event.eventId, ActionKind.CLOSE, output={"reason": requireNonBlank(reason, "关闭原因不能为空")})


def annotate(event: Event, *annotations: dict[str, Any]) -> Action:
    """附加解码树、标签或显示信息而不改变线上字节。"""
    return Action(event.eventId, ActionKind.ANNOTATE, annotations=list(annotations))


def redirect(event: Event, host: str, port: int) -> Action:
    """改写最终上游目标；主机为空或端口越界时直接报告作者输入错误。"""
    if isinstance(port, bool) or not isinstance(port, int) or port < 1 or port > 65535:
        raise ValueError("重定向目标必须包含有效主机和端口")
    return Action(event.eventId, ActionKind.REDIRECT, output={"host": requireNonBlank(host, "重定向主机不能为空"), "port": port})


def respond(event: Event, response: dict[str, Any]) -> Action:
    """生成完整合成响应；响应结构由对应 HTTP、DNS 或命令阶段定义。"""
    return Action(event.eventId, ActionKind.RESPOND, output=response)
